### PR TITLE
executor: support Chunk for UnionExec

### PR DESCRIPTION
**Pull request by zz-jason**
1. support `NextChunk` for UnionExec
2. for children of union, add projection to do type conversion if necessary.

### CODE REVIEW

File: executor/executor.go [hunk 1]
```diff
@@ -816,10 +866,32 @@ func (e *UnionExec) Next(goCtx goctx.Context) (Row, error) {
...
 }
 
+// NextChunk implements the Executor NextChunk interface.
+func (e *UnionExec) NextChunk(chk *chunk.Chunk) error {
+	chk.Reset()
+	if !e.initialized {
+		e.initialize(nil, false)
+		e.initialized = true
+	}
+	result, closed := <-e.resultPool
```
Comment (coocood) [inline executor/executor.go]: The second returned value is `ok`, not `closed`.

File: executor/executor.go [hunk 2]
```diff
@@ -767,23 +767,41 @@ func (e *MaxOneRowExec) Next(goCtx goctx.Context) (Row, error) {
...
-	wg       sync.WaitGroup
-	closedCh chan struct{}
+	stopFetchData atomic.Value
+	resultCh      chan *execResult
+	rows          []Row
+	cursor        int
+	wg            sync.WaitGroup
+
+	// For chunk execution.
+	resoucePool chan *executorResult
```
Comment (lamxTyler) [inline executor/executor.go]: `resouce` -> `resource`

File: executor/executor.go [hunk 3]
```diff
@@ -831,25 +849,58 @@ func (e *UnionExec) fetchData(goCtx goctx.Context, idx int) {
...
+}
+
+func (e *UnionExec) resultPuller(childID int) {
+	defer e.wg.Done()
+	for {
+		if e.stopFetchData.Load().(bool) {
+			return
+		}
+		resource := <-e.resoucePool
+		resource.err = errors.Trace(e.children[childID].NextChunk(resource.chk))
```
Comment (lamxTyler) [inline executor/executor.go]: Should we consider type conversion?

File: executor/executor.go [hunk 4]
```diff
@@ -869,10 +920,29 @@ func (e *UnionExec) Next(goCtx goctx.Context) (Row, error) {
...
+	if !ok {
+		return nil
+	}
+	if result.err != nil {
+		return errors.Trace(result.err)
+	}
+
+	chk.SwapColumns(result.chk)
+	result.reset()
+	e.resoucePool <- result
```
Comment (coocood) [inline executor/executor.go]: The resource pool may be closed by `waitAllFinished`.

How about just use a mutex protected slice for chunk reuse?

File: executor/executor.go [hunk 5]
```diff
@@ -767,23 +767,41 @@ func (e *MaxOneRowExec) Next(goCtx goctx.Context) (Row, error) {
...
 
 type execResult struct {
 	rows []Row
 	err  error
 }
 
-func (e *UnionExec) waitAllFinished() {
+type executorResult struct {
+	chk *chunk.Chunk
+	err error
```
Comment (coocood) [inline executor/executor.go]: How about adding a channel in executorResult to reuse the chunk and remove `resourcePool`?

File: executor/executor.go (hunk 3 above)
Comment (zz-jason) [reply executor/executor.go]: yes, we should

File: executor/executor.go (hunk 5 above)
Comment (zz-jason) [reply executor/executor.go]: I'll think about it

**Comment by zz-jason** [comment]
to #5261

File: executor/executor.go [hunk 6]
```diff
@@ -854,23 +854,36 @@ func (e *MaxOneRowExec) Next(goCtx goctx.Context) (Row, error) {
...
+	initialized    bool
+	chunkExecution bool
 }
 
 type execResult struct {
 	rows []Row
 	err  error
 }
 
+type executorResult struct {
```
Comment (XuHuaiyu) [inline executor/executor.go]: add comments for executorResult and execResult,
it's really similar.

File: executor/executor.go [hunk 7]
```diff
@@ -854,23 +854,36 @@ func (e *MaxOneRowExec) Next(goCtx goctx.Context) (Row, error) {
...
-	wg       sync.WaitGroup
-	closedCh chan struct{}
+	stopFetchData atomic.Value
+	resultCh      chan *execResult
+	rows          []Row
+	cursor        int
+	wg            sync.WaitGroup
+
+	// For chunk execution.
+	resourcePool   chan *executorResult
```
Comment (XuHuaiyu) [inline executor/executor.go]: add comments for resourcePool and resultPool

File: executor/executor.go [hunk 8]
```diff
@@ -918,25 +919,59 @@ func (e *UnionExec) fetchData(goCtx goctx.Context, idx int) {
...
+
+func (e *UnionExec) resultPuller(childID int) {
+	defer e.wg.Done()
+	for {
+		if e.stopFetchData.Load().(bool) {
+			return
+		}
+		resource := <-e.resourcePool
+		resource.err = errors.Trace(e.children[childID].NextChunk(resource.chk))
+		if resource.err != nil || resource.chk.NumRows() == 0 {
```
Comment (XuHuaiyu) [inline executor/executor.go]: We may not need nested loop.

File: executor/executor.go [hunk 9]
```diff
@@ -918,25 +919,59 @@ func (e *UnionExec) fetchData(goCtx goctx.Context, idx int) {
...
+			go e.fetchData(goCtx, i)
 		}
-		e.wg.Add(1)
-		go e.fetchData(goCtx, i)
 	}
 	go e.waitAllFinished()
-	return errors.Trace(err)
+}
+
+func (e *UnionExec) resultPuller(childID int) {
```
Comment (XuHuaiyu) [inline executor/executor.go]: I think it's better to add a comment for the execution flow of this func.

File: util/chunk/chunk.go [hunk 10]
```diff
@@ -92,6 +92,11 @@ func (c *Chunk) addColumnByFieldType(fieldTp *types.FieldType, initCap int) {
 	}
 }
 
+// SwapColumns swap columns with another Chunk.
```
Comment (XuHuaiyu) [inline util/chunk/chunk.go]: s/ swap/ swaps

File: executor/executor.go [hunk 11]
```diff
@@ -945,24 +973,42 @@ func (e *UnionExec) initialize(goCtx goctx.Context, forChunk bool) {
...
 		if e.stopFetchData.Load().(bool) {
 			return
 		}
-		resource := <-e.resourcePool
-		resource.err = errors.Trace(e.children[childID].NextChunk(resource.chk))
-		if resource.err != nil || resource.chk.NumRows() == 0 {
-			if resource.err != nil {
-				e.stopFetchData.Store(true)
+		select {
+		case _, ok := <-e.finished:
```
Comment (coocood) [inline executor/executor.go]: We don't need to check `ok`, it's always false, just return.

File: executor/executor.go [hunk 12]
```diff
@@ -945,24 +973,42 @@ func (e *UnionExec) initialize(goCtx goctx.Context, forChunk bool) {
...
+		result.err = errors.Trace(e.children[childID].NextChunk(result.chk))
+		if result.err == nil && result.chk.NumRows() == 0 {
+			return
+		}
+		select {
+		case _, ok := <-e.finished:
+			if !ok {
+				return
 			}
+		case e.resultPool <- result:
```
Comment (coocood) [inline executor/executor.go]: We can just send without `select`, the `resultPool` has enough capacity, it never blocks the sender.

**Comment by coocood** [comment]
LGTM

**Pull request merged**